lisa/tests/scheduler/uclamp.py

```diff
@@ -0,0 +1,209 @@
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import functools
+import itertools
+from operator import itemgetter
+import pandas as pd
+
+from lisa.analysis.frequency import FrequencyAnalysis
+from lisa.analysis.load_tracking import LoadTrackingAnalysis
+from lisa.datautils import df_add_delta
+from lisa.datautils import series_mean
+from lisa.tests.base import ResultBundle, RTATestBundle
+from lisa.wlgen.rta import Periodic
+
+
+class Uclamp(RTATestBundle):
+    """
+    Validate that UClamp min values are honored properly by the kernel.
+
+    The test is split into 8 phases. For each phase, a UClamp value is set for a
+    task, whom duty cycle would generate a lower utilization. Then the actual
+    capacity, allocated to the task during its activation is checked: The lower
+    boundary corresponds to the first CPU capacity available, bellow the UClamp
+    min. The upper boundary is 3 capacities above UClamp max, to take into
+    account the migration and schedutil margins.
+
+    The 8 phases UClamp values are picked to cover the entire SoC's CPU scale.
+    (usually between 0 and 1024)
+
+
+                 |<-- band 0 -->|<-- band 1 -->|<-- band 2 -->|<-- ...
+    capacities:  0      |      128     |      256            512
+                 |      |       |      |                      |
+    -------------|------|-------|------|----------------------|--------
+    phase 1:     |  uclamp_val  |      |                      |
+            expected_min        |      |                  expected_max
+                                |      |
+    ----------------------------|------|--------------------------------
+    phase 2:                    |  uclamp_val
+                           expected_min                             ...
+       ...
+
+    phase 8:
+    """
+
+    @classmethod
+    def _collect_capacities(cls, plat_info):
+        max_capacities = plat_info['cpu-capacities']['rtapp']
+
+        # returns, for each CPU a mapping frequency / capacity.
+        # cpu:
+        #   freq: capacity
+        #
+        # where capacity = max_cpu_capacity * freq / max_cpu_frequency.
+        return {
+                cpu.cpus[0]: {
+                                freq: int(max_capacities[cpu.cpus[0]] * freq /
+                                          max(cpu.active_states.keys()))
+                                for freq in cpu.active_states
+                }
+                for cpu in plat_info['nrg-model'].cpu_nodes
+        }
+
+    @classmethod
+    def _collect_capacities_flatten(cls, plat_info):
+        capacities = []
+
+        for cpu, states in cls._collect_capacities(plat_info).items():
+            for freq, cap in states.items():
+                capacities.append(cap)
+
+        return sorted(list(dict.fromkeys(capacities)))
```

Comment:
@valschneider I left a comment on the new version, I think it's better to go with a plain comprehension here, since I don't think it really brings anything to stack "shorthands" for the sake of saving a couple of lines (it's not like the problem was something cleanly solvable by let's say in pseudo code `sorted(map(itertools.chain.from_iterable^5, deeply_nested))`)